### ISSUE TITLE
Numerical Cancellation Issue

### ISSUE THREAD

**Issue by darioizzo** [darioizzo/audi#29]
Lets consider the simple computation:

```
import pyaudi as pa
x = pa.gdual_double(1e-10, "x", 2)
res = pa.sqrt(x)*pa.sqrt(x)
```
The exact result of this operation is ```x``` and hence the gdual representing it should be ```1e-10 + dx``` but, instead we get:

```
1e-10+dx-9.53674e-07*dx**2
```

This is an issue with numerical cancellation as explained now. 

Consider the polinomial ```sqrt(x+dx) = a+b[dx]+c[dx]**2``` and square it. The result is:

```
x+[dx] = a^2 + 2ab[dx] + (b^2+2ac)[dx]^2
```

Now, in the case of square root, we have that:

```
sqrt(x+dx) = 1e-5 + 50000[dx] -125000000000000[dx]^2
```
hence, in floating point arithmetic on my machine: ```b^2 = 2500000000.0``` and ```2*a*c = -2500000000.000001```

Thus, computing ```b^2+2ac``` gives a ```1E-7``` instead of zero.

The problem becomes more and more visible the closer you get to zero, or the higher the derivative.

Possible solutions are:

* go to multiple precision
* see if anything clever and generci can be done do face cancellation issues at the piranha level?

**Comment by bluescarni** [comment]
In practice with FP arithmetic you cannot rely on ``asin(sin(x)) == x`` or ``exp(log(x)) == x``, even with just plain scalar values. So I am not sure if it's worth it to coerce mathematically inexact computations to give mathematically exact answers.

With that said, I am having trouble analysing what is going on here exactly. If I compute ``b**2`` and ``2*a*c`` directly from C++, I get exactly the same value (both using ``std::pow()`` and just multiplying for squaring):

```c++
#include <cmath>
#include <iostream>

int main()
{
        std::cout.precision(40);
        std::cout << (std::pow(50000.,2.) - 2.*1E-5*125000000000000.)<< '\n';
        std::cout << (50000.*50000. - 2.*1E-5*125000000000000.) << '\n';
}
```

So I tried following what piranha should be doing in terms of term-by-term multiplications and additions, and it seems to me that the sequence of operations should be something like this:

```
1e-5 * 1e-5 -> constant term
1e-5 * 5e4 + 1e-5 * 5e4 -> linear term
1e-5 * 1.25e14 + 1e-5 * 1.25e14 - 5e4 * 5e4 -> quadratic term
```

But it still seems like everything is being calculated exactly. In particular, something like:

```
std::cout << (1e-5*1.25e14  - 5e4*5e4 + 1e-5*1.25e14) << '\n';
```

seems to always give zero regardless of how the operands are ordered (which makes sense since they have all the same magnitude, so I would not expect strange numerical effects).

So perhaps it would be helpful to understand what's going on step by step. The only thing that right now comes to my mind is that on new processors piranha might be using the ``std::fma()`` functions to do fused multiply-add of floating-point coefficient during polynomial multiplication. Perhaps that plays a role? Also, are you using any unsafe math optimisation when compiling?

**Comment by darioizzo** [comment]
On audi (python) the three coefficients are computed as:
```
a = 1e-05
b = 50000.0
c = -125000000000000.03
```
note that ```c``` has that .03 bullshit due, I guess, to FP shit and that will then explode into the cancellation problem.

**Comment by darioizzo** [comment]
Note also that this can get very sick. For example at order 40, already with dx = 0.1 we get errors of the magnitude of 1e+12 (yes +12)

**Comment by bluescarni** [comment]
Ah but then this happens in the computation of the ``sqrt()`` of the polynomial and not in the subsequent multiplication?

**Comment by darioizzo** [comment]
the .03 happens in the sqrt yes, but it has no consequence there. Then, that error, explodes when you do (for example) the function inversion or, in the case shown, ```sqrt(x)*sqrt(x)```

**Comment by bluescarni** [comment]
Sure, but the value in ``.03`` is not the exact result in the first place. Or is it?

**Comment by darioizzo** [comment]
No, it is not. Its a FP error made during the computation of the sqrt.

**Comment by bluescarni** [comment]
How is it ultimately computed? Directly with ``std::sqrt()`` or via some poly operation?

**Comment by darioizzo** [comment]
https://github.com/darioizzo/audi/blob/master/include/audi/functions.hpp#L241

You can look at this code for the cubic root. The square root will be similar.

It basically does piranha multiplications over polinomials and uses std::sqrt (std::cbrt) as a trigger.

**Comment by bluescarni** [comment]
It seems like the current sqrt code is using pow, which in turn invokes exp and log. Is that right? I would be inclined to say that there's some loss of precision due to the transcendent functions being used on floating-point values, but without a break down step by step of what's going on I cannot say it with 100% certainty.

Going multiprecision will help with the accuracy loss, how much you need will of course depend on how close to zero and how high the derivative. It's also gonna carry a heavy performance penalty.

**Comment by darioizzo** [comment]
No, thats why I pointed you to ```cbrt``` and not ```sqrt```.

The current implementation of ```sqrt``` is irrelevant. I use sqrt as an example as its easy to understand the cancellation. The issue is there identical with ```cbrt``` and other functions.

**Comment by darioizzo** [comment]
```
In [77]: x = pa.gdual_double(1e-10, "x", 2)

In [78]: pa.cbrt(x)*pa.cbrt(x)*pa.cbrt(x)
Out[78]: 1e-10+dx-7.15256e-07*dx**2
```

**Comment by bluescarni** [comment]
Is the cbrt code also losing information (i.e., producing non exact results in one of the coefficients of the polynomial produced by ``pa.cbrt(x)``)?

**Comment by darioizzo** [comment]
Its FP they are exact to 1e-14, then its rubbish.

**Comment by darioizzo** [comment]
In the case of cbrt:
```
a = 0.0004641588833612778
b = 1547196.2778709258
c = -5157320926236420.0
```

**Comment by bluescarni** [comment]
Yeah but that's what's happening, isn't it?

The fact that you have an error of 1E-7 on a quadratic term makes me think this is all coming down from that 1E-14 you lost in an inexact computation in the first place. How can you expect to recover the original exact result when there's lossy computation being done at every step? It might work in certain conditions/platforms if all you want is ``x == sqrt(x)*sqrt(x)``, but the moment you start intermixing other computations you are bound to accumulate some sort of error.

**Comment by darioizzo** [comment]
I agree. I do not think there is a solution. SOmetimes there may be tricks to perform the computation is some other way to get rid of the cancellation, but I am unaware of something that would work in general.

So the only think I can do at the moment is going mp, which will gain something but not eliminating the problem. (in sick case I loose 4-5 digits per order)

**Comment by bluescarni** [comment]
Isn't part of the issue here the functions as well? The fact that the derivatives of sqrt/cbrt go to infinite close to zero (and increasingly so the higher the differentiation order) seems like a recipe for numerical disasters, but maybe I am misunderstanding something.

**Comment by darioizzo** [comment]
Not sure. In the direct computation of the sqrt nothing wrong happens also at high order. You do get big numbers yes, and that probably cause the cancellation when trying to get a zero coefficient out of manipulating those huuge numbers.

**Comment by bluescarni** [comment]
btw before going full mp, maybe you want to try out with ``long double``. This should give you an idea of how the approach is working, ``long double`` has a couple extra decimal digits over ``double`` on x86 linux.

**Comment by darioizzo** [comment]
For sin infact the problem is not there:
```
In [107]: pa.sin(pa.asin(x)).find_cf([40])
Out[107]: -1.3438814711982046e-20

```

**Comment by bluescarni** [comment]
I tried to copy that into my pyaudi session and got a segfault :(

**Comment by bluescarni** [comment]
Full repro:

```ipython
In [2]: import pyaudi as pa

In [3]:  x = pa.gdual_double(0.5**33, "x", 2)

In [4]: pa.sin(pa.asin(x)).find_cf([40])
[1]    31261 illegal hardware instruction (core dumped)  ipython
```

**Comment by bluescarni** [comment]
This is pyaudi from pip, installed in local directory.

**Comment by darioizzo** [comment]
With the release of 1.2.4 the problem "31261 illegal hardware instruction (core dumped)  ipython" could be fixed by implementing your suggested changes in the compilation of gmp, can you check @bluescarni ?

**Comment by bluescarni** [comment]
Yep it works now:

```ipython
In [1]: import pyaudi as pa

In [2]: x = pa.gdual_double(0.5**33, "x", 50)

In [3]: x = pa.gdual_double(0.5**33, "x", 2)

In [4]: pa.sin(pa.asin(x)).find_cf([40])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-ddb06b840920> in <module>()
----> 1 pa.sin(pa.asin(x)).find_cf([40])

ValueError: requested coefficient is beyond the truncation order.

In [5]:
```

**Comment by darioizzo** [comment]
cool, one less problem. Will keep this open to post results on experiments using mp.
